assignments/burmanm/2.3.4.scm

```diff
@@ -0,0 +1,193 @@
+;; From SICP:
+
+(define (make-leaf symbol weight) (list 'leaf symbol weight))
+(define (leaf? object) (eq? (car object) 'leaf))
+(define (symbol-leaf x) (cadr x))
+(define (weight-leaf x) (caddr x))
+(define (make-code-tree left right)
+  (list left
+	right
+	(append (symbols left) (symbols right))
+	(+ (weight left) (weight right))))
+(define (left-branch
+	 tree) (car
+		tree))
+(define (right-branch tree) (cadr tree))
+(define (symbols tree)
+  (if (leaf? tree)
+      (list (symbol-leaf tree))
+      (caddr tree)))
+(define (weight tree)
+  (if (leaf? tree)
+      (weight-leaf tree)
+      (cadddr tree)))
+
+(define (decode bits tree)
+  (define (decode-1 bits current-branch)
+    (if (null? bits)
+	'()
+	(let ((next-branch
+	       (choose-branch (car bits) current-branch)))
+	  (if (leaf? next-branch)
+	      (cons (symbol-leaf next-branch)
+		    (decode-1 (cdr bits) tree))
+	      (decode-1 (cdr bits) next-branch)))))
+  (decode-1 bits tree))
+
+(define (choose-branch bit branch)
+  (cond ((= bit 0) (left-branch branch))
+	((= bit 1) (right-branch branch))
+	(else (error "bad bit: CHOOSE-BRANCH" bit))))
+
+(define (adjoin-set x set)
+  (cond ((null? set) (list x))
+	((< (weight x) (weight (car set))) (cons x set))
+	(else (cons (car set)
+		    (adjoin-set x (cdr set))))))
+
+(define (make-leaf-set pairs)
+  (if (null? pairs)
+      '()
+      (let ((pair (car pairs)))
+	(adjoin-set (make-leaf (car pair)
+			       (cadr pair))
+					; symbol
+					; frequency
+		    (make-leaf-set (cdr pairs))))))
+
+(define (append list1 list2)
+  (if (null? list1)
+      list2
+      (cons (car list1) (append (cdr list1) list2))))
+
+;; Exercise 2.67
+
+
+(define sample-tree
+  (make-code-tree (make-leaf 'A 4)
+		  (make-code-tree
+		   (make-leaf 'B 2)
+		   (make-code-tree
+		    (make-leaf 'D 1)
+		    (make-leaf 'C 1)))))
+(define sample-message '(0 1 1 0 0 1 0 1 0 1 1 1 0))
+
+;; scheme@(guile-user)> (decode sample-message sample-tree)
+;; $2 = (A D A B B C A)
+
+;; Exercise 2.68
+
+;; From SICP
+(define (encode message tree)
+  (if (null? message)
+      '()
+      (append (encode-symbol (car message) tree)
+	      (encode (cdr message) tree))))
+
+;;
+(define (encode-symbol symbol tree)
+  (define (encode-symbol-inner symbol tree bits)
+    (cond ((leaf? tree)
+	   (if (eq? (symbol-leaf tree) symbol)
+	       bits
+	       (error "No symbol found from the tree: " symbol)))
+	  ((eq? (symbol-leaf (left-branch tree)) symbol) (append bits '(0)))
```

Comment:
This doesn't seems to work when a left-branch is a tree. Though it works with the provided sample-tree, but I think you needs to use `(element-of-set? symbol (symbols (left-branch tree)))`